drivers/network/dd/e1000/hardware.c

```diff
@@ -696,15 +751,40 @@ NICGetPermanentMacAddress(
     USHORT AddrWord;
     UINT n;
 
+    ULONG Rah;
+    ULONG Ral;
+
     NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
 
-    /* Should we read from RAL/RAH first? */
-    for (n = 0; n < (IEEE_802_ADDR_LENGTH / 2); ++n)
+    /*
+     * First try to get the MAC via Ral/Rah. The first is read from NVM after a soft reset
+     * (which occurs in ndis.c just before calling this function)
+     */
+    E1000ReadUlong(Adapter, E1000_REG_RAH, &Rah);
+    if (Rah & E1000_RAH_AV)
+    {
+        NDIS_DbgPrint(MIN_TRACE, ("MAC from RAH/RAL seems to be useable...\n"));
+        Rah &= 0xffff;
+        E1000ReadUlong(Adapter, E1000_REG_RAL, &Ral);
+
+        *(ULONG *)Adapter->PermanentMacAddress = Ral;
+        *(USHORT *)&Adapter->PermanentMacAddress[4] = (USHORT)Rah;
+    }
+    else
     {
-        if (!E1000ReadEeprom(Adapter, (UCHAR)n, &AddrWord))
-            return NDIS_STATUS_FAILURE;
-        Adapter->PermanentMacAddress[n * 2 + 0] = AddrWord & 0xff;
-        Adapter->PermanentMacAddress[n * 2 + 1] = (AddrWord >> 8) & 0xff;
+        NDIS_DbgPrint(MIN_TRACE, ("Try to get the MAC by directly reading from EEPROM...\n"));
```

Comment:
The Linux e1000e driver, which I consider a trusted reference, does it exactly the other way round: First reading from the EEPROM, then from RAH. If there is a MAC address found in the EEPROM, it is even written into RAH to override any non-matching address. This is done to support pre-boot software, which programs an alternate MAC address into the EEPROM.

See https://elixir.bootlin.com/linux/v5.0/source/drivers/net/ethernet/intel/e1000e/e1000.h#L538 which calls a chip-specific read_mac_addr function, like https://elixir.bootlin.com/linux/v5.0/source/drivers/net/ethernet/intel/e1000e/82571.c#L1772 for the 8257x series of network adapters. This one calls into https://elixir.bootlin.com/linux/v5.0/source/drivers/net/ethernet/intel/e1000e/mac.c#L140 for reading the alternate MAC address.

As we cannot perform as much testing as the Linux driver has received and don't want to debug subtle bugs in the future, I highly suggest we follow their example and implement our driver in the same way.